tests/test_ssl.py

```diff
@@ -2997,6 +3095,225 @@ def test_wantWriteError(self) -> None:
+            )
+            logger.debug(f"Got SSL error: {e!r} ({reason}).")
+            return True  # Bad write retry
+
+    def _shutdown_connections(
+        self,
+        client: Connection,
+        server: Connection,
+        client_socket: socket,
+        server_socket: socket,
+    ) -> None:
+        """Helper to safely shut down SSL connections and close sockets."""
+        logger.debug("--- Cleanup: Shutting down connections ---")
+        try:
+            if client:
+                client.shutdown()
+        except Exception:
+            # An exception is usually thrown so it is caught and ignored.
+            pass
+        try:
+            if server:
+                server.shutdown()
+        except Exception:  # pragma: no cover
+            pass
+        finally:
+            if client_socket:
+                client_socket.close()
+            if server_socket:
+                server_socket.close()
+
+    def _badwriteretry(
+        self,
+        want_bad_retry: bool,
+        modeflag: int | None,
+    ) -> bool:
+        """
+        Tests for a "bad write retry" error over an SSL connection
+        by using a moving buffer which is allowed by default with
+        SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER. If this mode is unset
+        we expect a bad write retry error when using a moving buffer.
+        want_bad_retry: If True, the caller expects a bad write retry error.
+        mode: If not None, unsets the defaults that include
+        SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER and replaces with the given mode.
+        Returns True if a bad write retry error occurs.
+        """
+        request_buffer_size = 65536  # Size of the send buffer we'll request
+        client_socket, server_socket, client, server, sndbuf, rcvbuf = (
+            create_ssl_nonblocking_connection(modeflag, request_buffer_size)
+        )
+        result = False  # Default return value
+        # set buffer size to half the minimum of send and receive buffers
+        buffer_size = min(sndbuf, rcvbuf) // 2
+
+        # --- Main Test Flow ---
+        # _attempt_want_write_error() terminates the test
+        # if WantWriteError is not triggered
+        # The function also returns the message that triggered
+        # the WantWriteError so that when we attempt a retry
+        # we can ensure a different buffer location is allocated
+        # to a the new message we will send for the retry.
+        _ = self._attempt_want_write_error(client, buffer_size)
+
+        # proceed with draining so that a retry has a chance to succeed
+        self._drain_server_buffers(server, server_socket)
+
+        # now attempt the moving buffer retry
+        result = self._perform_moving_buffer_test(
+            client, buffer_size, want_bad_retry
+        )
+        self._shutdown_connections(
+            client, server, client_socket, server_socket
+        )
+        return result
```

Comment:
This looks like a sort of thing that seems natural to wrap with `@contextlib.contextmanager` or (even better `@pytest.fixture` with a `yield` for cleanup and indirect params).